Generic type declaration should be mandatory in function returning a generic type

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->
**V doctor:**
```
OS: linux, Ubuntu 18.04.5 LTS
Processor: 4 cpus, 64bit, little endian, Intel(R) Core(TM) i5-2400 CPU @ 3.10GHz
CC version: cc (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0

getwd: /Git/vlab
vmodules: /home/alejandro/.vmodules
vroot: /opt/v
vexe: /opt/v/v
vexe mtime: 2021-05-02 18:53:11
is vroot writable: true
is vmodules writable: true
V full version: V 0.2.2 273655e.fb7ddcd

Git version: git version 2.17.1
Git vroot status: weekly.2021.17-60-gfb7ddcd4
.git/config present: true
thirdparty/tcc status: thirdparty-linux-amd64 7543de81
```

**What did you do?**
`v -g -o vdbg cmd/v && vdbg example3.v`
```v
struct GenericChannelStruct<T> {
	ch chan T
}

struct Simple {
	msg string
}

fn main() {
	d := new_channel_struct<Simple>()
}

pub fn new_channel_struct<T>() GenericChannelStruct {
	d := GenericChannelStruct{
		ch: chan T{}
	}

	return d
}

```

**What did you expect to see?**

I think a compiler error, as GenericChannelStruct is a generic type, but "new_channel_struct" function return type doesn't carry the generic type.

**What did you see instead?**
```
example3.v:10:2: warning: unused variable: `d`
    8 | 
    9 | fn main() {
   10 |     d := new_channel_struct<Simple>()
      |     ^
   11 | }
   12 |
==================
/tmp/v/example3.18348186199058710381.tmp.c:1991: error: ';' expected (got "main__new_channel_struct_T_main__Simple")
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

If you were not working with C interop, this is a compiler bug, please report the bug using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

```

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -6794,6 +6794,15 @@ fn (mut c Checker) fn_decl(mut node ast.FnDecl) {
 			c.error('only functions that do NOT return values can have `[if $ct_name]` tags',
 				node.pos)
 		}
+		if node.generic_names.len > 0 {
+			gs := c.table.get_type_symbol(node.return_type)
+			if gs.info is ast.Struct {
+				if gs.info.is_generic && !node.return_type.has_flag(.generic) {
+					c.error('return generic struct in fn declaration must specify the generic type names, e.g. Foo<T>',
+						node.return_type_pos)
+				}
+			}
+		}
 	}
 	if node.is_method {
 		mut sym := c.table.get_type_symbol(node.receiver.typ)

--- a/vlib/v/checker/tests/generics_fn_return_generic_struct_err.out
+++ b/vlib/v/checker/tests/generics_fn_return_generic_struct_err.out
@@ -1,0 +1,7 @@
+vlib/v/checker/tests/generics_fn_return_generic_struct_err.vv:13:32: error: return generic struct in fn declaration must specify the generic type names, e.g. Foo<T>
+   11 | }
+   12 |
+   13 | pub fn new_channel_struct<T>() GenericChannelStruct {
+      |                                ~~~~~~~~~~~~~~~~~~~~
+   14 |     d := GenericChannelStruct{
+   15 |         ch: chan T{}

--- a/vlib/v/checker/tests/generics_fn_return_generic_struct_err.vv
+++ b/vlib/v/checker/tests/generics_fn_return_generic_struct_err.vv
@@ -1,0 +1,18 @@
+struct GenericChannelStruct<T> {
+	ch chan T
+}
+
+struct Simple {
+	msg string
+}
+
+fn main() {
+	new_channel_struct<Simple>()
+}
+
+pub fn new_channel_struct<T>() GenericChannelStruct {
+	d := GenericChannelStruct{
+		ch: chan T{}
+	}
+	return d
+}

